nexpose/nexpose.py

```diff
@@ -2052,6 +2062,33 @@ def GenerateScanReport(self, scan_or_id):
         body = ''.join(data[4:-1])
         return as_xml(base64.urlsafe_b64decode(body))
 
+    def GenerateAdHocReport(self, adhoc_report_configuration):
+        """
+        Generate adhoc report and return decoded contents.
+        """
+        # TODO: add optional filename param to store the report to disk instead of memory
+        self._RequireInstanceOf(adhoc_report_configuration, AdhocReportConfiguration)
+        data = self.RequestAdhocReport(adhoc_report_configuration.AsXML())
+        data = self.VerifySuccess(data)
+        # TODO: figure out a way to handle this safely/correctly for different formats
+        data = data.tail.replace('\r', '').strip().split('\n')
+        # assert data[1] == 'Content-Type: text/xml; name=report.xml'
+        assert data[2] == 'Content-Transfer-Encoding: base64'
+        assert data[3] == ''
+        assert data[0] == data[-1][:-2]
+        body = ''.join(data[4:-1])
```

Comment:
I had experience when requesting an XML report & within the response there were multiple report items.
i.e. the response would look like this:

```
<ReportAdhocGenerateResponse success="1"/>
--AxB9sl3299asdjvbA
Content-Type: image/gif; name=Chart00000001.gif
Content-Transfer-Encoding: base64
<Base64 encoded gif>
--AxB9sl3299asdjvbA
Content-Type: image/gif; name=Chart00000005.gif
Content-Transfer-Encoding: base64
<Base64 encoded gif>
--AxB9sl3299asdjvbA
Content-Type: text/xml; name=report.xml
Content-Transfer-Encoding: base64
<Base64 encoded xml>
--AxB9sl3299asdjvbA
Content-Type: image/gif; name=Chart00000002.gif
Content-Transfer-Encoding: base64
<Base64 encoded gif>
```

So in this case, there should be some looping trying to find the actual report.

Also, it's really confusing to see splits & then calling each element without an example response.

Reply:
Yeah this is just copied from the existing adhoc report method (which only takes a scan id) and it seems pretty limited. I would like this to be able to work the way the ruby gem does where you can write to file instead of holding it all in memory. Being able to capture the extra attachments (like charts) would be nice, but the ruby gem doesn't support that either so I'm not super concerned about it.

Reply:
I agree with this. After experimenting a bit, it is possible to use Python's [MIME parser](https://docs.python.org/3/library/email.parser.html?highlight=message_from_string#email.message_from_string) to handle these cases but you need to modify the input slightly, i.e. prefix some headers. Here is an excerpt of how we implemented it:

```
        full_msg = 'Content-Type: {}\r\n\r\n{}'.format(
            resp.headers['Content-Type'],
            resp.text,
        )
        msg = email.message_from_string(full_msg)
        if not msg.is_multipart():
            raise RuntimeError("Got unexpected (non multipart MIME) response from GenerateAdhocReport: {}".format(msg))


        vuln_data = None
        # There will be several parts in the response (images etc.),
        # so linear search for the part that is the actual report.
        for part in msg.walk():
            if part.get_filename() == "report.csv":
                assert part.get_content_type() == 'text/csv'

                # Decode the bytes.
                csv_ = part.get_payload(decode=True).decode('utf8')
                vuln_data = csv.DictReader(io.StringIO(csv_))
                break
        else:
            raise RuntimeError("Could not find report in GenerateAdhocReport response")
```

A similar approach could work here as well I assume. That way all the verification can be deferred to a library but the content needs to be prepared properly.